portal/migrations/versions/5caf794c70a7_.py

```diff
@@ -0,0 +1,125 @@
+    return patients_query(
+        acting_user=admin,
+        include_test_role=False,
+        filter_by_ids=[i[0] for i in at_least_once_wd],
+        requested_orgs=irnmn_orgs).with_entities(User.id)
+
+
+def audit_since_for_patient(patient):
+    #version_pattern = '24.3.8'
+    min_id = 864688  # last id on prod with version 23.12.11.2
+    # ignore `assessment` audits, as that's primarily qb_timeline rebuilds
+    # which all of these users had.
+    audit_query = Audit.query.filter(Audit._context != 'assessment').filter(
+        Audit.subject_id == patient.id).filter(Audit.id > min_id)
+    if not audit_query.count():
+        return
+    for audit in audit_query:
+        if audit.context == 'consent' and audit.comment.startswith('remove bogus'):
+            continue
+        # TODO counts from audit depending on interaction
+        print(audit)
+
+def confirm_withdrawn_row(patient, rs_id):
+    """confirm a withdrawal row is in users qb_timeline for given research study"""
+    count = QBT.query.filter(QBT.user_id == patient.id).filter(
+        QBT.research_study_id == rs_id).filter(QBT.status == 'withdrawn').count()
+    if count != 1:
+        raise ValueError(f"no (or too many) withdrawn row for {patient.id} {rs_id}")
+
+
+def mask_withdrawn_user_email(session, patient, admin):
+    if not patient.email_ready()[0]:
+        # no need, already hidden
+        return
+
+    version = current_app.config.metadata['version']
+    now = datetime.datetime.utcnow()
+    def audit_insert(subject_id):
+        msg = f"mask withdrawn user email"
+        insert = (
+            "INSERT INTO AUDIT"
+            " (user_id, subject_id, context, timestamp, version, comment)"
+            " VALUES"
+            f"({admin.id}, {subject_id}, 'demographics',"
+            f" '{now}', '{version}', '{msg}')")
+        session.execute(insert)
+
+    patient._email = WITHDRAWN_PREFIX + patient.email
+    audit_insert(patient.id)
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = Session(bind=bind)
+    admin = session.query(User).filter_by(email='__system__').first()
+
+    for pat_row in patients(admin):
+        patient_id = pat_row[0]
+        patient = session.query(User).get(patient_id)
+        # confirm withdrawn status (global g or empro e)
+        consent_g, withdrawal_g = consent_withdrawal_dates(patient, 0)
+        consent_e, withdrawal_e = consent_withdrawal_dates(patient, 1)
+        if not (withdrawal_e or withdrawal_g):
+            continue
+
+        if withdrawal_g:
+            confirm_withdrawn_row(patient, 0)
+        if withdrawal_e:
+            confirm_withdrawn_row(patient, 1)
+
+        # check for users consented for both but only withdrawn from one
+        if (consent_g and consent_e) and not (withdrawal_g and withdrawal_e):
+            print(f"user {patient_id} consented for both, but only withdrawn from one")
+            continue
+
+        mask_withdrawn_user_email(session, patient, admin)
```

Comment:
@pbugni  in slack you wrote "[12 hours ago](https://cirg.slack.com/archives/C5MH9NVKQ/p1711431751401259?thread_ts=1711382681.633649&cid=C5MH9NVKQ)
_one hack that would buy more time: i could write a tiny migration to alter the email address of all withdrawn users, so we’d be certain to not send anything to them.  this would be reversible, and enable us to turn the job back on._"

My read was that you were going to do this out of band as a temporary stopgap. There is the possibility that consent will be restored for some of these - the system needs to back this out if that happens.

Reply:
@mcjustin yes - please see commit 9e46f28 below.  now applying the mask (and removing it) on all consent changes when applicable.